cli/src/ui.rs

```diff
@@ -566,15 +566,15 @@ impl Ui {
                 .unwrap_or(false)
     }
 
-    pub fn prompt(&self, prompt: &str) -> io::Result<String> {
+    pub fn prompt(&self, writer: &mut dyn Write, prompt: &str) -> io::Result<String> {
         if !Self::can_prompt() {
```

Comment:
Can you elaborate the intent? It's weird that `can_prompt()` tests `stdout`, but we use an arbitrary write stream.

If the goal was to handle the case where `stdout` is redirected, it might be okay to always use `stderr`. Maybe we can also use some library that can obtain tty even if stdout/err was redirected. I assume `rpassword` does something like that.

Reply:
I assume its for testing Google's custom binary. But I agree that its missing the intent in the commit message.

Reply:
Updated the commit description to be clearer, and refactored the change to fix the disconnect between `can_prompt()` and the actual output channel.

The internal use case I'm trying to support here is running `jj` in a shell function environment where we need to parse `stdout` and can't have it polluted with other messaging.  It's not really possible to use some sort of escaping mechanism to work around the problem of `prompt` printing to stdout because then either the user doesn't see the prompt message (if we hide stdout), or else we print the parsed output the user isn't supposed to see in addition to the prompt message (if we don't hide stdout).

Reply:
Thanks. Maybe we can just always use `stderr` then? I don't think there is a reason we have to use `stdout`.

Reply:
Sure, it's certainly an easier change.